src/librustc/infer/error_reporting/nice_region_error/placeholder_error.rs

```diff
@@ -311,13 +314,46 @@ impl NiceRegionError<'me, 'gcx, 'tcx> {
         sup_placeholder: Option<ty::Region<'tcx>>,
         has_sub: Option<usize>,
         has_sup: Option<usize>,
-        expected_trait_ref: ty::TraitRef<'_>,
-        actual_trait_ref: ty::TraitRef<'_>,
+        expected_trait_ref: ty::TraitRef<'tcx>,
+        actual_trait_ref: ty::TraitRef<'tcx>,
         vid: Option<ty::Region<'tcx>>,
         expected_has_vid: Option<usize>,
         actual_has_vid: Option<usize>,
         any_self_ty_has_vid: bool,
     ) {
+        // HACK(eddyb) maybe move this in a more central location.
+        #[derive(Copy, Clone)]
+        struct Highlighted<'a, 'gcx, 'tcx, T> {
+            tcx: TyCtxt<'a, 'gcx, 'tcx>,
+            highlight: RegionHighlightMode,
+            value: T,
+        }
+
+        impl<'a, 'gcx, 'tcx, T> Highlighted<'a, 'gcx, 'tcx, T> {
+            fn map<U>(self, f: impl FnOnce(T) -> U) -> Highlighted<'a, 'gcx, 'tcx, U> {
+                Highlighted {
+                    tcx: self.tcx,
+                    highlight: self.highlight,
+                    value: f(self.value),
+                }
+            }
+        }
+
+        impl<'a, 'gcx, 'tcx, T> fmt::Display for Highlighted<'a, 'gcx, 'tcx, T>
+            where T: for<'b, 'c> Print<'gcx, 'tcx,
+                FmtPrinter<'a, 'gcx, 'tcx, &'b mut fmt::Formatter<'c>>,
+                Error = fmt::Error,
+            >,
+        {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Comment:
Something about this impl doesn't *feel right* to me. I mean, I'm not saying it's buggy or anything like that. I guess it's fine. I think I'm just a bit sad about how this Display impl creates the pretty-printer, rather than adding a bit of configuration to an existing one...somehow... 

Basically, I was sort of hoping we can get to a place where you can kind of "compose" flags like this, so that you could write:

```
foo.with_highlight(...).with_highlight(...).with_fuzzy_bear(...)
```

and thus print `foo` with all those things. Or..something like that. Anyway, I don't think that's necessarily something to address here, it's just something I'd like to think about maybe in some future PR.

Reply:
I want to take configuration out of `FmtPrinter` (and do it in types instead of data, basically).

The reason the printer is created here is that *this is a printing entry-point*.

That is, `Highlighted<T>` display-formats similar to `T` does, but with different configuration.

The `T` wouldn't have been `ty::print::Print::print`-ed without going through `fmt::Display`, so there's no meaningful "existing printer".

An earlier version of this code created a pretty-printer once but when someone touched the diagnostics I had to introduce this wrapper type mid-rebase.